signature-aggregator/aggregator/aggregator.go

```diff
@@ -179,6 +178,115 @@ func (s *SignatureAggregator) connectToQuorumValidators(
 	return connectedValidators, nil
 }
 
+// getUnderfundedL1Nodes fetches the set of L1 nodes that are underfunded
+// It uses the underfundedL1NodeCache to avoid fetching the same data multiple times
+func (s *SignatureAggregator) getUnderfundedL1Nodes(
+	ctx context.Context,
+	log logging.Logger,
+	signingSubnet ids.ID,
+	skipCache bool,
+) (set.Set[ids.NodeID], error) {
+	fetchUnderfundedL1Nodes := func(subnetID ids.ID) (set.Set[ids.NodeID], error) {
+		validators, err := s.pChainClient.GetCurrentValidators(ctx, subnetID, nil, s.pChainClientOptions...)
+		if err != nil {
+			log.Error(
+				"Failed to fetch current L1 validators",
+				zap.String("subnetID", subnetID.String()),
+				zap.Error(err),
+			)
+			return nil, fmt.Errorf("failed to fetch current L1 validators: %w", err)
+		}
+
+		underfundedL1Nodes := set.NewSet[ids.NodeID](0)
+		for _, v := range validators {
+			if v.ClientL1Validator.ValidationID == nil {
+				log.Debug(
+					"Skipping non-L1 validator",
+					zap.String("nodeID", v.NodeID.String()),
+				)
+				continue
+			}
+
+			l1Validator := v.ClientL1Validator
+
+			if l1Validator.Balance == nil {
+				underfundedL1Nodes.Add(v.NodeID)
+				log.Warn(
+					"Node has nil balance",
+					zap.String("nodeID", v.NodeID.String()),
+				)
+				continue
+			}
+
+			if *l1Validator.Balance < minimumL1ValidatorBalance {
+				underfundedL1Nodes.Add(v.NodeID)
+				log.Debug(
+					"Node has insufficient balance",
+					zap.String("nodeID", v.NodeID.String()),
+					zap.Uint64("balance", *l1Validator.Balance),
+					zap.Any("signingSubnetID", signingSubnet),
+				)
+			}
+		}
+
+		return underfundedL1Nodes, nil
+	}
+
+	underfundedL1Nodes, err := s.underfundedL1NodeCache.Get(signingSubnet, fetchUnderfundedL1Nodes, skipCache)
+	if err != nil {
+		log.Error(
+			"Failed to get underfunded L1 nodes",
+			zap.String("signingSubnetID", signingSubnet.String()),
+			zap.Error(err),
+		)
+		return nil, err
+	}
+
+	return underfundedL1Nodes, nil
+}
+
+func (s *SignatureAggregator) getExcludedValidators(
```

Comment:
Would it make more sense to cache the set of `excluded` validators instead of the set of `underfunded` validators?

Reply:
Ah, I see we get the Validator set from a different API. That makes sense then :)